### ISSUE TITLE
PromptTuning and PTuning with SEQ_2_SEQ_LM

### ISSUE THREAD

**Issue by VukVukovic** [huggingface/peft#386]
I'm trying to use Parameter-Efficient Fine-tuning approaches for a `SEQ_2_SEQ_LM` tasks (`ought/raft`: `twitter_complaints`, `tweet_eval_hate`). I'm mainly testing with models such as `google/flan-t5-base`, `bigscience/T0_3B`, `google/t5-base-lm-adapt`. I am using a dataset tokenized with model's tokenizer, DataCollatorForSeq2Seq, as well as a regular train/eval loops. I'm getting very good results (both metrics as well as the decoded sequence) when using PrefixTuning, however, I cannot get results any close to these when using PromptTuning or PTuning.

I've tried several changes to mitigate these problems, however, the problem still persists.

PromptTuning:
- Using `num_transformer_submodules=1`

PTuning, PromptTuning:
- Removing `num_virtual_tokens` tokens from the left side of the model outputs during evaluations (as mentioned in [Caveats](https://github.com/huggingface/peft#caveats) 2.)

Is there anything else I'm missing?

**Comment by fotinidelig** [comment]
Hi! I have an additional question to your experiments: aren't you getting a NotImplementedError due to [this line](https://github.com/huggingface/peft/blob/main/src/peft/peft_model.py#L970) with PromptTuning?

I have the same issue when using Prompt Tuning on the google/t5-small-lm-adapt model, task_type='CausalLM', and get really bad results.

Perhaps there is a deeper problem here w/ encoder-decoder models pre-trained on mask problems?

**Comment by VukVukovic** [comment]
Hi @fotinidelig! I was getting NotImplementedError when decoding, however, that's why I used logits of the classification labels from `outputs = model(**inputs)`.

> I have the same issue when using Prompt Tuning on the google/t5-small-lm-adapt model, task_type='CausalLM', and get really bad results.
That sounds like the same problem.

> Perhaps there is a deeper problem here w/ encoder-decoder models pre-trained on mask problems?
It's just unclear to me how PrefixTuning works properly.

**Comment by meettyj** [comment]
It seems that the promptTuning does not work with "google/flan-t5-small". The following code block will raise NotImplementedError. Any comments or thoughts?

```
from peft import get_peft_model
from transformers import AutoModelForSeq2SeqLM, AutoTokenizer
from peft import LoraConfig, TaskType, PromptTuningConfig, PromptTuningInit
import torch

model_name_or_path = "google/flan-t5-small"
tokenizer_name_or_path = "google/flan-t5-small"
model = AutoModelForSeq2SeqLM.from_pretrained(model_name_or_path)

peft_config = PromptTuningConfig(
    task_type=TaskType.SEQ_2_SEQ_LM,
    prompt_tuning_init=PromptTuningInit.TEXT,
    num_virtual_tokens=8,
    prompt_tuning_init_text="Choose the best option to answer the question:",
    tokenizer_name_or_path=model_name_or_path,
)


model = get_peft_model(model, peft_config)
model.print_trainable_parameters()

tokenizer = AutoTokenizer.from_pretrained(tokenizer_name_or_path)

device = 'cuda:0'
model = model.to(device)
model.eval()
inputs = tokenizer("Tweet text : @HondaCustSvc Your customer service has been horrible during the recall process. I will never purchase a Honda again. Label :", return_tensors="pt")["input_ids"]
print('inputs: ', inputs.shape)

with torch.no_grad():
    outputs = model.generate(input_ids=inputs.to(device))
    print(tokenizer.batch_decode(outputs.detach().cpu().numpy(), skip_special_tokens=True)[0])

```

Thank you!

**Comment by StefanHeng** [comment]
> 



> Hi! I have an additional question to your experiments: aren't you getting a NotImplementedError due to [this line](https://github.com/huggingface/peft/blob/main/src/peft/peft_model.py#L970) with PromptTuning?
> 
> I have the same issue when using Prompt Tuning on the google/t5-small-lm-adapt model, task_type='CausalLM', and get really bad results.
> 
> Perhaps there is a deeper problem here w/ encoder-decoder models pre-trained on mask problems?

I have the same issue. I don't think `generate` is supported for `seq2seqLM` w/ `p tuning` and `prompt tuning`. I also mentioned this issue [here](https://github.com/huggingface/peft/pull/543#issuecomment-1581156157). **When do the team plan to support this feature**? 

Pinging first two contributors of this repo. @pacman100 @younesbelkada

**Comment by pacman100** [comment]
Hello everyone, PR https://github.com/huggingface/peft/pull/519 resolves this issue
